File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;
        int res = 0;
        int leftMax = height[l];
        int rightMax = height[r];
        
        while(l < r) {
            if(leftMax < rightMax) {
                l++;
                leftMax = max(height[l], leftMax);
                res = res + (leftMax - height[l]);
            } else {
                r--;
                rightMax = max(height[r], rightMax);
                res = res + (rightMax - height[r]);
            }
        }
        return res;

    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        left_max = height[l]
        right_max = height[r]
        res = 0
        while l < r:
            left_max = max(height[l], left_max)
            right_max = max(height[r], right_max)
            res += (left_max - height[l]) + (right_max - height[r])
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return res